### ISSUE TITLE
[Feature] Can we only save a part of model parameters?

### ISSUE THREAD

**Issue by hmtbgc** [open-mmlab/mmengine#1294]
### What is the feature?

In some complicated models(such as MiniGPT-4), there are several different modules that are parameter-frozen when training. It is not necessary to save the whole parameters for each validation. However, in mmengine, it seems that we cannot only save a part of model parameters( e.g. parameters that require gradient)

In mmengine/runner/runner.py, there is a function named 'save_checkpoint':
```python        
checkpoint = {
      'meta':meta,
      'state_dict':weights_to_cpu(model.state_dict()),
      'message_hub':apply_to(self.message_hub.state_dict(),
               lambda x: hasattr(x, 'cpu'), lambda x: x.cpu()),
  }
```

maybe we can first select parameters and then save?
``` python
param_grad_dict = {
            k: v.requires_grad for (k, v) in model.named_parameters()
        }
state_dict = model.state_dict()
for k in list(state_dict.keys()):
  if k in param_grad_dict.keys() and not param_grad_dict[k]:
      del state_dict[k]
checkpoint = {
      'meta':meta,
      'state_dict':weights_to_cpu(state_dict),
      'message_hub':apply_to(self.message_hub.state_dict(),
               lambda x: hasattr(x, 'cpu'), lambda x: x.cpu()),
  }
```

This could reduce the total saved paramters from 30G to 126M (under MiniGPT-4 setting)

Or can we let users choose whether save total parameters or just a part?

**Comment by zhouzaida** [comment]
Hi @hmtbgc , you can implement a `state_dict` in your model and it will be called by `save_checkpoint`.

https://github.com/open-mmlab/mmengine/blob/398d229910f02dbddfd4659ee53d5b570ea6a14c/mmengine/runner/runner.py#L2201

**Comment by takfate** [comment]
Reimplement the `state_dict` in the root module

**Comment by hmtbgc** [comment]
Thanks, i got it.
